release/ray_release/byod/build.py

```diff
@@ -2,51 +2,75 @@
 
 import boto3
 import hashlib
+import os
 import subprocess
 import sys
+import time
 
 from ray_release.logger import logger
 from ray_release.test import Test
 
 DATAPLANE_S3_BUCKET = "ray-release-automation-results"
 DATAPLANE_FILENAME = "dataplane.tgz"
 DATAPLANE_DIGEST = "f9b0055085690ddad2faa804bb6b38addbcf345b9166f2204928a7ece1c8a39b"
+BASE_IMAGE_WAIT_TIMEOUT = 7200
+BASE_IMAGE_WAIT_DURATION = 30
 
 
 def build_anyscale_byod_images(tests: List[Test]) -> None:
     """
     Builds the Anyscale BYOD images for the given tests.
     """
     _download_dataplane_build_file()
+    to_be_built = {}
     built = set()
     for test in tests:
         if not test.is_byod_cluster():
             continue
-        ray_image = test.get_ray_image()
-        if ray_image in built:
-            continue
-        byod_image = test.get_anyscale_byod_image()
-        logger.info(f"Building {byod_image} from {ray_image}")
-        with open(DATAPLANE_FILENAME, "rb") as build_context:
-            subprocess.check_call(
-                [
-                    "docker",
-                    "build",
-                    "--build-arg",
-                    f"BASE_IMAGE={ray_image}",
-                    "-t",
-                    byod_image,
-                    "-",
-                ],
-                stdin=build_context,
-                stdout=sys.stderr,
-                env={"DOCKER_BUILDKIT": "1"},
-            )
-            subprocess.check_call(
-                ["docker", "push", byod_image],
-                stdout=subprocess.DEVNULL,
-            )
-            built.add(ray_image)
+        to_be_built[test.get_ray_image()] = test
+
+    env = os.environ.copy()
+    env["DOCKER_BUILDKIT"] = "1"
+    timeout = BASE_IMAGE_WAIT_TIMEOUT
+    # ray images are built on post-merge, so we can wait for them to be available
+    while len(built) < len(to_be_built) and timeout > 0:
+        for ray_image, test in to_be_built.items():
+            byod_image = test.get_anyscale_byod_image()
+            if _byod_image_exist(test):
+                logger.info(f"Image {byod_image} already exists")
+                built.add(ray_image)
+                continue
+            if not _ray_image_exist(ray_image):
+                # TODO(can): instead of waiting for the base image to be built, we can
+                #  build it ourselves
+                logger.info(
+                    f"Image {ray_image} does not exist yet. "
+                    f"Wait for another {timeout}s..."
+                )
+                time.sleep(BASE_IMAGE_WAIT_DURATION)
+                timeout -= BASE_IMAGE_WAIT_DURATION
```

Comment:
this will be kind of not accurate if `_ray_image_exist()` can hang. it is better to track the time out based on the starting time, and time elapsed, rather than tracking the total amount of sleep time.